programs/keeper/Keeper.cpp

```diff
@@ -288,20 +288,25 @@ int Keeper::main(const std::vector<std::string> & /*args*/)
     /// Check that the process user id matches the owner of the data.
     const auto effective_user_id = geteuid();
     struct stat statbuf;
-    if (stat(path.c_str(), &statbuf) == 0 && effective_user_id != statbuf.st_uid)
+    const auto effective_user = getUserName(effective_user_id);
```

Comment:
By the way you may not have any mapping for uid to login, so I guess it is better to uid into error message too.

Reply:
yes, we have this situation. It seems to use uid for effective_user if there is no username.

Reply:
I still don't see UIDs in the error message.

Reply:
What do you see, is it blank? 
For example, this is what was printing when some of the tests were failing:
```
Effective user of the process (root) does not match the owner (1000) of a data 
```
In this case I believe user 1000 (`ubuntu` in the host) is not in passwd and prints `1000`. This is what the old implementation did as well.  As this code was there originally, so you should been seeing it with the original implementation as well?

Reply:
If I understand correctly `getpwuid_r` doesn't return an error if the password entry doesn't exist but rather returns no `result`. The code seems to be working on that basis...

```cpp
    const auto error = getpwuid_r(user_id, &passwd_entry, buffer.data(), buffer_size, &result);

    if (error)
        throwFromErrno("Failed to find user name for " + toString(user_id), ErrorCodes::FAILED_TO_GETPWUID, error);
    else if (result)
        return result->pw_name;
    return toString(user_id);
````

Reply:
Yes, you are right, I haven't looked at `getUserName()`, thanks.